plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php

```diff
@@ -138,6 +138,12 @@ public function create( &$product ) {
 		if ( $id && ! is_wp_error( $id ) ) {
 			$product->set_id( $id );
 
+			// get the post object so that we can set the status
+			// to the correct value; it is possible that the status was
+			// changed by the woocommerce_new_product_data filter above.
+			$post_object = get_post( $product->get_id() );
+			$product->set_status( $post_object->post_status );
+
```

Comment:
I was tempted to just call `$this->read( $product );` after the call to `$product->apply_changes()` below, to ensure that everything about the product reflects what was stored in the DB. But, that is a much bigger change in my opinion, has more room for subtle backwards-compatibility issues, and might have more of a performance impact. I still think that change should be made, but that is not immediately needed (while this is, for #48194).

Reply:
Gathering some thoughts on this, not immediately sure what is best. An existing pattern used elsewhere, though, basically takes this form:

```php
$product = new WC_Product;
$product->save();
$product = wc_get_product( $product->get_id() );
```

I'm not suggesting that is ideal, or what most would expect to have to do, though.